src/layers/icmp.rs

```diff
@@ -0,0 +1,107 @@
+//! ICMP Datagram
+
+use std::convert::TryInto;
+
+use serde::Serialize;
+
+use crate::errors::Error;
+use crate::layers::ipv4;
+use crate::types::IPv4Address;
+use crate::Layer;
+
+/// IANA Assigned protocol number for ICMP
+pub const IPPROTO_ICMP: u8 = 1_u8;
+
+// Register ICMP with Protocol Handler in IPv4
+pub(crate) fn register_defaults() -> Result<(), Error> {
+    ipv4::register_protocol(IPPROTO_ICMP, ICMP::creator)
+}
+
+#[derive(Default, Debug, Serialize)]
+#[serde(untagged)]
+pub enum IcmpType {
+    #[default]
+    Unused,
+    Echo(IcmpEcho),
+    Redirect(IcmpRedirect),
+}
+
+#[derive(Default, Debug, Serialize)]
+pub struct IcmpEcho {
+    identifier: u16,
+    sequence_number: u16,
+}
+
+#[derive(Default, Debug, Serialize)]
+pub struct IcmpRedirect {
+    gateway_address: IPv4Address,
+}
+
+/// Structure representing the ICMP Header
+#[derive(Default, Debug, Serialize)]
+pub struct ICMP {
+    #[serde(rename = "type")]
+    icmp_type: u8,
+    code: u8,
+    checksum: u16,
+    #[serde(flatten)]
+    rest_of_header: IcmpType,
+}
+
+impl ICMP {
+    pub(crate) fn creator() -> Box<dyn Layer + Send> {
+        Box::<ICMP>::default()
+    }
+}
+
+impl Layer for ICMP {
+    fn decode_bytes(
+        &mut self,
+        bytes: &[u8],
+    ) -> Result<(Option<Box<dyn Layer + Send>>, usize), Error> {
+        let decoded;
+
+        if bytes.len() < 8 {
```

Comment:
This probably should be a `const` like `ICMP_HEADER_LENGTH` or something like that? Improves readability.